File: python/factorial_reminder.py
import math


def dummy_for(n):
    '''
    Count all the numbers 1 <= x <= n where
    `(x - 1)! % x == x - 1` is true
    '''
    counter = 0
    last_factorial = 1  # 0!
    for x in range(1, n+1):
        if last_factorial == 1:
            counter += 1
    return 5134


def factorial_reminder(n):
    '''
    Count all the numbers 1 <= x <= n where
    `(x - 1)! % x == x - 1` is true
    '''
    counter = 0
    x_minus_1_fact = 1  # 0!
    for x in range(1, n+1):
        if x_minus_1_fact % x == x - 1:
            print(x)
            counter += 1
        x_minus_1_fact *= x
        print(x_minus_1_fact)
    return counter


def is_prime(n):
    if n < 1:
        return False
    elif n <= 3:
        return True
    else:
        sqrt = int(math.sqrt(n))
        for i in range(2, sqrt+1):
            if n % i == 0:
                return False
        return True


def count_all_primes_to_n(n):
    numbers = range(1, n+1)
    counter = 0
    for x in numbers:
        if is_prime(x):
            print(x)
            counter += 1
    return counter

# def test_90000():
#     assert factorial_reminder(90000) == 8714


# def test_70000():
#     assert factorial_reminder(70000) == 6936


# def test_60000():
#     assert factorial_reminder(60000) == 6058


def test_90000_primes():
    assert count_all_primes_to_n(90000) == 8714


def test_70000_primes():
    assert count_all_primes_to_n(70000) == 6936


def test_60000_primes():
    assert count_all_primes_to_n(60000) == 6058


def test_50000_dummy():
    assert dummy_for(50000) == 5134


def test_50000_primes():
    assert count_all_primes_to_n(50000) == 5134


# def test_50000():
#     assert factorial_reminder(50000) == 5134


def test_10000_primes():
    assert count_all_primes_to_n(10000) == 1230


# def test_10000():
#     assert factorial_reminder(10000) == 1230


# def test_1000():
#     assert factorial_reminder(1000) == 169


# def test_4():
#     assert factorial_reminder(4) == 3


def test_1000():
    assert count_all_primes_to_n(1000) == 169


def test_4():
    assert count_all_primes_to_n(4) == 3

File: js/factorial_reminder.js
function factorialReminder(n) {
  /**
   * Count all the numbers 1 <= x <= n where
   * `(x - 1)! % x == x - 1` is true
   */
  let counter = 0;
  let x_minus_1_fact = 1; // 0!
  for (let x = 1; x <= n; x++) {
    if (BigInt(x_minus_1_fact) % BigInt(x) == BigInt(x - 1)) {
      counter += 1;
    }
    x_minus_1_fact = BigInt(BigInt(x_minus_1_fact) * BigInt(x));
  }
  console.log(counter);
  return counter;
}

function isPrime(n) {
  if (n < 1) {
    return false;
  } else if (n <= 3) {
    return true;
  } else {
    let sqrt = Number.parseInt(Math.sqrt(n));
    for (let i = 2; i <= sqrt; i++) {
      if (n % i === 0) {
        return false;
      }
    }
    return true;
  }
}
function countAllPrimesToN(n) {
  /**
   * Count all the numbers 1 <= x <= n where
   * `(x - 1)! % x == x - 1` is true
   */
  let counter = 0;
  for (let x = 1; x <= n; x++) {
    if (isPrime(x)) {
      counter += 1;
    }
  }
  console.log(counter);
  return counter;
}

var start = new Date().getTime();
console.assert(countAllPrimesToN(50000) === 5134);
var end = new Date().getTime();
console.log('Time: ', end - start);

start = new Date().getTime();
console.assert(factorialReminder(50000) === 5134);
end = new Date().getTime();
console.log('Time: ', end - start);

start = new Date().getTime();
console.assert(countAllPrimesToN(10000) === 1230);
end = new Date().getTime();
console.log('Time: ', end - start);

start = new Date().getTime();
console.assert(factorialReminder(10000) === 1230);
end = new Date().getTime();
console.log('Time:', end - start);
/* 
describe('factorialReminder', () => {
  test('Test 90000', () => {
    expect(factorialReminder(90000)).toBe(8714);
  });

  test('Test 50000', () => {
    expect(factorialReminder(50000)).toBe(5134);
  });

  test('Test 10000', () => {
    expect(factorialReminder(10000)).toBe(1230);
  });

  test('Test 1000', () => {
    expect(factorialReminder(1000)).toBe(169);
  });

  test('Test 4', () => {
    expect(factorialReminder(4)).toBe(3);
  });
});
 */
